spec/services/email_service/ses/synchronize_spec.rb

```diff
@@ -0,0 +1,125 @@
+
+  def gen_address(type)
+    status, verification_requested_at = get_type(type).call()
+
+    { id: next_id(),
+      community_id: 1,
+      name: gen_str(25),
+      email: gen_email(),
+      verification_status: status,
+      verification_requested_at: verification_requested_at }
+  end
+
+  def ids(addresses)
+    addresses.map { |a| a[:id] }
+  end
+
+  def emails(addresses)
+    addresses.map { |a| a[:email] }
+  end
+
+  def store_addresses(addresses)
+    addresses.map { |addr|
+      m = MarketplaceSenderEmail.new(addr)
+      m.id = addr[:id]
+      m.save!
+      m.reload
+    }
+  end
+
+  describe "#build_sync_updates" do
+    it "marks verified requested as verified" do
+      addresses = 10.times.map { gen_address(:new_pending) }
+      verified_addresses = emails(addresses[0...5])
+
+      expect(Synchronize.build_sync_updates(addresses, verified_addresses))
+        .to eq({verified: ids(addresses[0...5]),
+                expired: [],
+                touch: ids(addresses[5...10]),
+                none: []})
+    end
+
+    it "marks old pending requested as expired" do
+      addresses = 10.times.map { gen_address(:old_pending) }
+      verified_addresses = emails(addresses[5...10])
+
+      expect(Synchronize.build_sync_updates(addresses, verified_addresses))
+        .to eq({verified: ids(addresses[5...10]),
+                expired: ids(addresses[0...5]),
+                touch: [],
+                none: []})
+
+    end
+
+    it "marks non-verified verified as none" do
+      addresses = 10.times.map { gen_address(:verified) }
+      verified_addresses = emails(addresses[5...10])
+
+      expect(Synchronize.build_sync_updates(addresses, verified_addresses))
+        .to eq({touch: ids(addresses[5...10]),
+                none: ids(addresses[0...5]),
+                verified: [],
+                expired: []})
+
+    end
+  end
+
+  describe "#run_synchronization" do
+    it "syncs db contents with SES data" do
+      now = Time.now()
+      Timecop.freeze(now) do
+
+        addresses = (Synchronize::BATCH_SIZE + 10).times.map { gen_address(:new_pending) }
+        verified_addresses = emails(addresses[0...Synchronize::BATCH_SIZE / 2])
+
+        original_addresses = store_addresses(addresses)
+        orig_updated_at_max = original_addresses.reduce(DateTime.new(0)) { |max, a| a[:updated_at] > max ? a[:updated_at]  : max }
+
+        stubs = {list_verified_email_addresses: {verified_email_addresses: verified_addresses}}
+        ses_client = EmailService::SES::Client.new(config: {region: "fake-region", access_key_id: "access_key", secret_access_key: "secret_access_key"},
+                                                   stubs: {list_verified_email_addresses: {verified_email_addresses: verified_addresses}})
+
+
+        Timecop.travel(now + 5.seconds) do
+          Synchronize.run_synchronization!(ses_client)
+
+          verified_addresses = MarketplaceSenderEmail.where(verification_status: "verified").pluck(:id)
```

Comment:
I feel that we instead of accessing the Model object, we could use some higher abstraction, like Store or even API.